sdks/java/core/src/main/java/org/apache/beam/sdk/io/LocalResourceId.java

```diff
@@ -0,0 +1,129 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.base.Preconditions.checkState;
+
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Objects;
+import java.util.UUID;
+import org.apache.beam.sdk.io.fs.ResolveOptions;
+import org.apache.beam.sdk.io.fs.ResolveOptions.StandardResolveOptions;
+import org.apache.beam.sdk.io.fs.ResourceId;
+import org.apache.commons.lang3.SystemUtils;
+
+/**
+ * {@link ResourceId} implementation for local files.
+ */
+class LocalResourceId implements ResourceId {
+
+  private final Path path;
+  private final boolean isDirectory;
+
+  static LocalResourceId fromPath(Path path, boolean isDirectory) {
+    checkNotNull(path, "path");
+    return new LocalResourceId(path, isDirectory);
+  }
+
+  private LocalResourceId(Path path, boolean isDirectory) {
+    this.path = path.normalize();
+    this.isDirectory = isDirectory;
+  }
+
+  @Override
+  public ResourceId resolve(String other, ResolveOptions resolveOptions) {
+    checkState(
+        isDirectory,
+        String.format("Expected the path is a directory, but had [%s].", path));
+    checkArgument(
+        resolveOptions.equals(StandardResolveOptions.RESOLVE_FILE)
+            || resolveOptions.equals(StandardResolveOptions.RESOLVE_DIRECTORY),
+        String.format("ResolveOptions: [%s] is not supported.", resolveOptions));
+    checkArgument(
+        !(resolveOptions.equals(StandardResolveOptions.RESOLVE_FILE)
+            && other.endsWith("/")),
+        "The resolved file: [%s] should not end with '/'.", other);
+    if (SystemUtils.IS_OS_WINDOWS) {
+      return resolveLocalPathWindowsOS(other, resolveOptions);
+    } else {
+      return resolveLocalPath(other, resolveOptions);
+    }
+  }
+
+  @Override
+  public ResourceId getCurrentDirectory() {
+    if (isDirectory) {
+      return this;
+    } else {
+      Path parent = path.getParent();
+      checkState(
+          parent != null,
+          String.format("Path: [%s] doesn't have the current directory.", path));
```

Comment:
This is when it is a file, but don't have the parent directory. such as the case in testGetCurrentDirectoryInvalid().